projects/concurrency/README.md

```diff
@@ -0,0 +1,308 @@
+}
+
+func (s *server) recordResult(isSuccess bool) {
+    s.mu.Lock()
+    defer s.mu.Unlock()
+    if isSuccess {
+        successfulRequests += 1
+    } else {
+        failedRequests += 1
+    }
+}
+
+func (s *server) handleRequest() {
+    s.mu.Lock()
+    totalRequests += 1
+    s.mu.Unlock()
+    recordResult(true)
+}
+```
+
+And this code would cause a deadlock, because one thread would be trying to acquire a lock it already holds:
+
+```go
+type server struct {
+    mu sync.Mutex
+}
+
+func (s *server) recordResult(isSuccess bool) {
+    s.mu.Lock()
+    defer s.mu.Unlock()
+    if isSuccess {
+        successfulRequests += 1
+    } else {
+        failedRequests += 1
+    }
+}
+
+func (s *server) handleRequest() {
+    s.mu.Lock()
+    defer s.mu.Unlock()
+    totalRequests += 1
+    recordResult(true)
+}
+```
+
+One pattern that's often used is to have some functions require a lock is held in order to call them - something like:
+
+```go
+type server struct {
+    mu sync.Mutex
+}
+
+// s.mu must be held to call this function.
+// The compiler won't guarantee this is the case, but hopefully developers will see the name and the comment and do the right thing.
+func (s *server) recordResult_locked(isSuccess bool) {
+    if isSuccess {
+        successfulRequests += 1
+    } else {
+        failedRequests += 1
+    }
+}
+
+func (s *server) handleRequest() {
+    s.mu.Lock()
+    defer s.mu.Unlock()
+    totalRequests += 1
+    recordResult_locked(true)
+}
+```
+
+### Project: LRU Cache with accurate statistics
+
+You are going to implement a cache.
+
+A cache is used to store some result (probably because it was in some way expensive to compute/fetch) so that it can be looked up more cheaply in the future.
+
+Caches (like many problems) involves trade-offs. We are using some memory (by storing results) to save some other resource in the future (often compute time).
+
+(Note that caches can get _very_ sophisticated - we're going to write quite a simple one).
+
+Often times, when we use a cache, we want to limit how much memory it will use. This means we need to sometimes delete things from the cache. There are different policies for cache eviction. The one we are going to pick is "least recently used" (or LRU).
+
+Imagine we have a limit of 3 items in our cache, and we already have three items in it. If we try to add an additional item, first, we need to remove one. We need to sort the entries by when they were most recently added or accessed, and remove the oldest.
+
+We also want to keep statistics about our cache. We want to be able to ask the following questions:
+1. How many entries were written to the cache and have never been read after being written (including for things which have been evicted)?
+2. What is the average number of times a cache entry is read (just for things currently in the cache)?
+3. How many reads and writes have been performed in the cache across all time (including for things which have been evicted)?
```

Comment:
hit rates are a huge thing we care about that's missing here.
how many reads can be served from cache?
would probably have a 'hits' and 'misses' counter and then one's monitoring can do rates over that and compute %s over time